### PR TITLE
BUG: Raise TypeError only if key DataFrame is not empty #10126

### PR DESCRIPTION
**Pull request by rekcahpassyla**
Proposed fix for #10126

### CODE REVIEW

**Comment by jreback** [comment]
change this to close issue #10126 instead

File: pandas/tests/test_frame.py [hunk 1]
```diff
@@ -794,6 +794,24 @@ def test_setitem_empty(self):
         result.loc[result.b.isnull(), 'a'] = result.a
         assert_frame_equal(result, df)
 
+    def test_setitem_empty_frame_with_boolean(self):
+        #
```
Comment (jreback) [inline pandas/tests/test_frame.py]: add the issue number here as a comment

File: pandas/tests/test_frame.py [hunk 2]
```diff
@@ -794,6 +794,24 @@ def test_setitem_empty(self):
...
         assert_frame_equal(result, df)
 
+    def test_setitem_empty_frame_with_boolean(self):
+        # 
+        df = pd.DataFrame([[2, 2]], index=pd.DatetimeIndex(['2011-01-01']))
+        df2 = pd.DataFrame([[1, 1]], index=pd.DatetimeIndex(['2011-01-01']))
+
+        df[df > df2] = 47
+
+        assert np.all(df.values==47)
```
Comment (jreback) [inline pandas/tests/test_frame.py]: construct an expected frame and use ``assert_frame_equal`` instead

File: pandas/core/frame.py [hunk 3]
```diff
@@ -2151,7 +2151,7 @@ def _setitem_array(self, key, value):
     def _setitem_frame(self, key, value):
         # support boolean setting with DataFrame input, e.g.
         # df[df > df2] = 0
-        if key.values.dtype != np.bool_:
+        if key.values.size and key.values.dtype != np.bool_:
```
Comment (jreback) [inline pandas/core/frame.py]: use ``is_bool_dtype`` here

**Comment by rekcahpassyla** [comment]
@jreback Requested changes made and squashed, changed title to reference #10126

File: pandas/tests/test_frame.py [hunk 4]
```diff
@@ -794,6 +794,27 @@ def test_setitem_empty(self):
         result.loc[result.b.isnull(), 'a'] = result.a
         assert_frame_equal(result, df)
 
+    def test_setitem_empty_frame_with_boolean(self):
+        # Test for issue #10126
+        df = pd.DataFrame([[2, 2]], index=pd.DatetimeIndex(['2011-01-01']))
+        df2 = pd.DataFrame([[1, 1]], index=pd.DatetimeIndex(['2011-01-01']))
+
+        df[df > df2] = 47
```
Comment (jreback) [inline pandas/tests/test_frame.py]: you don't need the first part of this test (e.g. before the loop), its just confusing.

File: pandas/tests/test_frame.py [hunk 5]
```diff
@@ -794,6 +794,27 @@ def test_setitem_empty(self):
...
+
+        expected = pd.DataFrame(
+            [[47, 47]], index=pd.DatetimeIndex(['2011-01-01'])
+        )
+        assert_frame_equal(df, expected)
+
+        for dtype in ('float', 'int64'):
```
Comment (jreback) [inline pandas/tests/test_frame.py]: leave from here on

**Comment by rekcahpassyla** [comment]
Using `is_bool_dtype` results in a couple of extra test failures- investigating.

```python
======================================================================
FAIL: test_setitem_boolean (pandas.tests.test_frame.TestDataFrame)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\dev\code\opensource\pandas-rekcahpassyla\pandas\tests\test_frame.py", line 621, in test_setitem_boolean
    df[df * 0] = 2
  File "c:\dev\code\opensource\pandas-rekcahpassyla\pandas\util\testing.py", line 1640, in __exit__
    raise AssertionError("{0} not raised.".format(name))
AssertionError: TypeError not raised.

======================================================================
FAIL: test_frame_getitem_setitem_boolean (pandas.tests.test_multilevel.TestMultiLevel)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\dev\code\opensource\pandas-rekcahpassyla\pandas\tests\test_multilevel.py", line 354, in test_frame_getitem_setitem_boolean
    df[df * 0] = 2
  File "c:\dev\code\opensource\pandas-rekcahpassyla\pandas\util\testing.py", line 1640, in __exit__
    raise AssertionError("{0} not raised.".format(name))
AssertionError: TypeError not raised.
```

**Comment by rekcahpassyla** [comment]
Fixed my obvious silly mistake. (Doh)

File: pandas/tests/test_frame.py [hunk 6]
```diff
@@ -794,6 +794,18 @@ def test_setitem_empty(self):
...
+            for df in [
+                    pd.DataFrame(dtype=dtype),
+                    pd.DataFrame(dtype=dtype, index=[1]),
+                    pd.DataFrame(dtype=dtype, columns=['A']),
+            ]:
+                df2 = df.copy()
+                df[df > df2] = 47
+
```
Comment (jreback) [inline pandas/tests/test_frame.py]: you still need an ``assert_frame_equal`` here to make that the assignment is correct.

**Comment by jreback** [comment]
ok, looks good. pls add a release mentioning the original issue.

**Comment by rekcahpassyla** [comment]
Rebased and added release note to 0.16.2, in the bug fix section.

**Comment by shoyer** [comment]
thanks!

**Comment by rekcahpassyla** [comment]
Thank you!

**Comment by jorisvandenbossche** [comment]
Since this is merged in master, all travis builds seem to fail (although the tests for the PR itself were green):

```
FAIL: test_sum (pandas.tests.test_series.TestSeries)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/tests/test_series.py", line 2319, in test_sum
    self._check_stat_op('sum', np.sum)
  File "/home/travis/build/pydata/pandas/pandas/tests/test_series.py", line 2683, in _check_stat_op
    testit()
  File "/home/travis/build/pydata/pandas/pandas/tests/test_series.py", line 2656, in testit
    self.assertTrue(np.isnan(f(allna)))
AssertionError: False is not true
```

**Comment by jorisvandenbossche** [comment]
Ah, caused by bottleneck upgrade: #10270

**Pull request merged**